src/cpp/session/modules/SessionPackages.R

```diff
@@ -283,63 +283,120 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
 
 .rs.addFunction("listInstalledPackages", function()
 {
-   # calculate unique libpaths
-   uniqueLibPaths <- .rs.uniqueLibraryPaths()
-
-   # get packages
-   x <- suppressWarnings(library(lib.loc=uniqueLibPaths))
-   x <- x$results[x$results[, 1] != "base", , drop=FALSE]
-   
-   # extract/compute required fields 
-   pkgs.name <- x[, 1]
-   pkgs.library <- x[, 2]
-   pkgs.desc <- x[, 3]
-   pkgs.url <- file.path("help/library",
-                         pkgs.name, 
-                         "html", 
-                         "00Index.html")
-   loaded.pkgs <- .rs.pathPackage()
-   pkgs.loaded <- !is.na(match(normalizePath(
-                                  paste(pkgs.library,pkgs.name, sep="/")),
-                               loaded.pkgs))
-   
-
-   # build up vector of package versions
-   instPkgs <- as.data.frame(installed.packages(), stringsAsFactors=F)
-   pkgs.version <- sapply(seq_along(pkgs.name), function(i){
-     .rs.packageVersion(pkgs.name[[i]], pkgs.library[[i]], instPkgs)
-   })
-
-   pkgs.browse <- sapply(seq_along(pkgs.name), function(i){
-     repo <- getOption("repos")[[1]]
-     if (!identical(repo, NULL)) {
-       slash <- if (.rs.lastCharacterIs(repo, "/")) "" else "/"
-       .rs.scalar(paste(repo, slash, "package=", pkgs.name[[i]], sep = ""))
-     } 
-     else {
-       NULL
-     }
+   # get the CRAN repository URL, and remove a trailing slash if required
+   repos <- getOption("repos")
+   cran <- if ("CRAN" %in% names(repos))
+      gsub("/*$", "", repos[["CRAN"]])
+   else
+      "https://cran.rstudio.com"
+   
+   # helper function for extracting information from a package's
+   # DESCRIPTION file
+   readPackageInfo <- function(pkgPath) {
+      
+      # attempt to read package metadata
+      desc <- .rs.tryCatch({
+         metapath <- file.path(pkgPath, "Meta", "package.rds")
+         metadata <- readRDS(metapath)
+         as.list(metadata$DESCRIPTION)
+      })
+      
+      # if that failed, try reading the DESCRIPTION
+      if (inherits(desc, "error"))
+         desc <- read.dcf(file.path(pkgPath, "DESCRIPTION"), all = TRUE)
+      
+      # attempt to infer an appropriate URL for this package
+      if (identical(as.character(desc$Priority), "base")) {
+         source <- "Base"
+         url <- ""
+      } else if (!is.null(desc$URL)) {
+         source <- "Custom"
+         url <- strsplit(desc$URL, "\\s*,\\s*")[[1]][[1]]
+      } else if ("biocViews" %in% names(desc)) {
+         source <- "Bioconductor"
+         url <- sprintf("https://www.bioconductor.org/packages/release/bioc/html/%s.html", desc$Package)
+      } else if (identical(desc$Repository, "CRAN")) {
+         source <- "CRAN"
+         url <- sprintf("%s/package=%s", cran, desc$Package)
+      } else if (!is.null(desc$GithubRepo)) {
+         source <- "GitHub"
+         url <- sprintf("https://github.com/%s/%s", desc$GithubUsername, desc$GithubRepo)
```

Comment:
Is URL encoding necessary on these fields? (same w/Bioconductor above)

Reply:
It normally shouldn't be, but in theory a package with malicious fields here could attempt to inject HTML that should be escaped, so I'm doing so now.